fuel-poa-coordinator/src/service.rs

```diff
@@ -0,0 +1,321 @@
+            None
+        }
+    }
+}
+
+pub struct Task<S>
+where
+    S: InterpreterStorage + Send,
+    <S as InterpreterStorage>::DataError: Send,
+{
+    stop: mpsc::Receiver<()>,
+    block_gas_limit: Word,
+    db: S,
+    block_producer: Arc<dyn BlockProducer>,
+    txpool_sender: TxPoolSender,
+    txpool_broadcast: broadcast::Receiver<TxStatusBroadcast>,
+    import_block_events_tx: broadcast::Sender<ImportBlockBroadcast>,
+    /// Last block creation time. When starting up, this is initialized
+    /// to `Instant::now()`, which delays the first block on startup for
+    /// a bit, but doesn't cause any other issues.
+    last_block_created: Instant,
+    trigger: Trigger,
+    /// Deadline clock, used by the triggers
+    timer: DeadlineClock,
+}
+impl<S> Task<S>
+where
+    S: InterpreterStorage + Send,
+    <S as InterpreterStorage>::DataError: Send,
+{
+    // Request the block producer to make a new block, and return it when ready
+    async fn signal_produce_block(&mut self) -> anyhow::Result<FuelBlock> {
+        let current_height = self
+            .db
+            .block_height()
+            .map_err(|err| anyhow::format_err!("db error {err:?}"))?;
+        let height = BlockHeight::from(current_height + 1);
+
+        self.block_producer
+            .produce_block(height, self.block_gas_limit)
+            .await
+    }
+
+    async fn produce_block(&mut self) -> anyhow::Result<()> {
+        // Ask the block producer to create the block
+        let block = self.signal_produce_block().await?;
+
+        // Send the block back to the txpool
+        // TODO: this probably must be done differently with multi-node configuration
+        self.import_block_events_tx
+            .send(ImportBlockBroadcast::PendingFuelBlockImported {
+                block: Arc::new(block),
+            })
+            .expect("Failed to import the generated block");
+
+        // Update last block time
+        self.last_block_created = Instant::now();
+
+        // Set timer for the next block
+        match self.trigger {
+            Trigger::Never => {
+                unreachable!("This mode will never produce blocks");
+            }
+            Trigger::Instant => {}
+            Trigger::Interval { block_time } => {
+                // TODO: instead of sleeping for `block_time`, subtract the time we used for processing
+                self.timer.set_timeout(block_time, OnConflict::Min).await;
+            }
+            Trigger::Hybrid {
+                max_block_time,
+                min_block_time,
+                max_tx_idle_time,
+            } => {
+                self.timer
+                    .set_timeout(max_block_time, OnConflict::Min)
+                    .await;
+
+                let block_tx_count =
+                    self.signal_produce_block().await?.transactions.len();
+                let includable_txs = self.txpool_sender.includable().await?.len();
```

Comment:
nit: Just for the sake of consistency: 

```suggestion
                let block_tx_count =
                    self.signal_produce_block().await?.transactions.len();
                let includable_txs_count = self.txpool_sender.includable().await?.len();
```

Reply:
Also, `includable()` returns the sorted txs. If you're only getting the count of the txs in the tx pool, and the order doesn't matter, you could get away with some sort of `size()` call on the `txpool_sender`, but that currently doesn't exist.

I would suggest maybe investigating to see if we can do a similar optimization to get the upcoming size of the block from the block producer without going through all the production steps.

Reply:
On thxpool side, we are sending a vector of `Arc`'d transactions and then taking count. The complier might be able to optimize this to just sending then len over.